platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java

```diff
@@ -425,93 +456,198 @@ public DataFileReader endWriting() throws IOException {
+            } catch (final IOException z) {
+                exceptionOccurred.set(z);
+                // Make sure the writing thread is resumed
+                finishWritingLatch.countDown();
+                completeExceptionally(z);
+                return false;
+            }
+        }
+    }
+
+    /**
+     * An "update" bucket task. Task input is a bucket (provided by "read bucket" task) and
+     * a list of updates to apply to the bucket. Has one dependency: the bucket to update
+     */
+    public class UpdateBucketTask extends AbstractTask {
+
+        // A bucket to update
+        private Bucket bucket;
+
+        // List of updates to apply to the bucket
+        private final BucketMutation keyUpdates;
+
+        UpdateBucketTask(final ForkJoinPool pool, final BucketMutation keyUpdates) {
+            super(pool, 1);
+            this.keyUpdates = keyUpdates;
+        }
+
+        void setBucket(final Bucket bucket) {
+            this.bucket = bucket;
+            send();
+        }
+
+        @Override
+        protected boolean exec() {
+            try {
+                keyUpdates.forEachKeyValue(bucket::putValue);
+                // Create a subsequent "store bucket" task for the bucket
+                final StoreBucketTask storeTask = new StoreBucketTask(getPool(), bucket);
+                // The last created "store bucket" task. storeTask above will be set as an
+                // output dependency for that task to make sure tasks are running only one at
+                // a time. See StoreBucketTask for details
+                final StoreBucketTask prevTask = lastStoreTask.getAndSet(storeTask);
+                if (prevTask != null) {
+                    // This will trigger prevTask execution as soon as its prev task is complete
+                    prevTask.setNext(storeTask);
+                } else {
+                    // The first task: no dependency on the prev task, can be executed rightaway
+                    storeTask.send();
+                }
+                if (storeBucketTasksCreated.incrementAndGet() == numBuckets.get()) {
+                    // The last task: no dependency on the next task, can be executed as soon as
+                    // its prev task is complete, no need to wait until the next task dependency
+                    // is set
+                    lastStoreTask.get().send();
+                }
+                return true;
+            } catch (final Exception z) {
+                exceptionOccurred.set(z);
+                // Make sure the writing thread is resumed
+                finishWritingLatch.countDown();
+                completeExceptionally(z);
+                return false;
+            }
+        }
+    }
+
+    public class StoreBucketTask extends AbstractTask {
+
+        private final Bucket bucket;
+
+        private StoreBucketTask next;
+
+        StoreBucketTask(final ForkJoinPool pool, final Bucket bucket) {
+            super(pool, 2);
```

Comment:
It's worth clarifying that this configuration will result in sequential execution of these tasks.

Reply:
Oh, I was going to provide some JavaDoc for `StoreBucketTask`, but forgot about it... I'll describe this task and its dependencies shortly